### DISCUSSION TITLE
Map Render Blanks

### DISCUSSION THREAD

**Discussion by firdauskz01** [mapsforge/vtm#854] [Q&A] [answered]
Hi, I'm trying to compile on Android SDK 29, but I got white blank. This my code :
```
 UrlTileSource tileSource = DefaultSources.OPENSTREETMAP.build();

            // Vector layer
            VectorTileLayer tileLayer = mapView.map().setBaseMap(tileSource);

            // Building layer
            mapView.map().layers().add(new BuildingLayer(mapView.map(), tileLayer));

            // Label layer
            mapView.map().layers().add(new LabelLayer(mapView.map(), tileLayer));

            // Render theme
            mapView.map().setTheme(VtmThemes.DEFAULT);

            // Scale bar
            MapScaleBar mapScaleBar = new DefaultMapScaleBar(mapView.map());
            MapScaleBarLayer mapScaleBarLayer = new MapScaleBarLayer(mapView.map(), mapScaleBar);
            mapScaleBarLayer.getRenderer().setPosition(GLViewport.Position.BOTTOM_LEFT);
            mapScaleBarLayer.getRenderer().setOffset(5 * CanvasAdapter.getScale(), 0);
            mapView.map().layers().add(mapScaleBarLayer);

            // Note: this map position is specific to Berlin area
            mapView.map().setMapPosition(52.517037, 13.38886, 1 << 12);
            mapView.map().render();
            mapView.map().updateMap();
```

and this is my manifest

```
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="mapsforgevtm.test">

    <uses-permission android:name="android.permission.INTERNET" />
    <application
        android:allowBackup="true"
        android:icon="@mipmap/ic_launcher"
        android:label="@string/app_name"
        android:roundIcon="@mipmap/ic_launcher_round"
        android:supportsRtl="true"
        android:requestLegacyExternalStorage="true"
        android:usesCleartextTraffic="true"
        android:theme="@style/Theme.MapsforgeVTM">
        <activity android:name=".MainActivity">
            <intent-filter>
                <action android:name="android.intent.action.MAIN" />

                <category android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
        </activity>
    </application>

</manifest>
```

and this is my app.build gradle
```plugins {
    id 'com.android.application'
}

android {
    compileSdkVersion 29
    buildToolsVersion "30.0.2"

    defaultConfig {
        applicationId "mapsforgevtm.test"
        minSdkVersion 21
        targetSdkVersion 29
        versionCode 1
        versionName "1.0"

        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
    }

    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
        }
    }
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
}

dependencies {

    implementation 'androidx.appcompat:appcompat:1.3.0'
    implementation 'com.google.android.material:material:1.3.0'
    implementation 'androidx.constraintlayout:constraintlayout:2.0.4'
    testImplementation 'junit:junit:4.13.2'
    androidTestImplementation 'androidx.test.ext:junit:1.1.2'
    androidTestImplementation 'androidx.test.espresso:espresso-core:3.3.0'

    //vtm core
    implementation 'org.mapsforge:vtm:0.16.0'
    implementation 'org.mapsforge:vtm-themes:0.16.0'
    implementation 'net.sf.kxml:kxml2:2.3.0'
    implementation 'org.slf4j:slf4j-api:1.7.29'
    
//    vtm android libgdx
    implementation 'org.mapsforge:vtm-android:0.16.0'
    implementation 'org.mapsforge:vtm-android:0.16.0:natives-armeabi-v7a'
    implementation 'org.mapsforge:vtm-android:0.16.0:natives-arm64-v8a'
    implementation 'org.mapsforge:vtm-android:0.16.0:natives-x86'
    implementation 'org.mapsforge:vtm-android:0.16.0:natives-x86_64'
    implementation 'org.mapsforge:vtm-gdx:0.16.0'
    implementation 'org.mapsforge:vtm-android-gdx:0.16.0'
    implementation 'org.mapsforge:vtm-android-gdx:0.16.0:natives-armeabi-v7a'
    implementation 'org.mapsforge:vtm-android-gdx:0.16.0:natives-arm64-v8a'
    implementation 'org.mapsforge:vtm-android-gdx:0.16.0:natives-x86'
    implementation 'org.mapsforge:vtm-android-gdx:0.16.0:natives-x86_64'
    implementation 'com.badlogicgames.gdx:gdx:1.9.10'
    implementation 'com.badlogicgames.gdx:gdx-backend-android:1.9.10'
    implementation 'com.caverock:androidsvg:1.4'
}
```

this is the result

![image](https://user-images.githubusercontent.com/19215383/122733726-a3897c80-d2a7-11eb-9128-70e3d332c94a.png)

perhaps I'm missing something? Please help me

**Comment by devemux86** [comment, marked as answer]
See the [getting started example](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/GettingStarted.java) (and [integration guide](https://github.com/mapsforge/vtm/blob/master/docs/Integration.md)).

And compare your code with the various samples in [vtm-android-example](https://github.com/mapsforge/vtm/tree/master/vtm-android-example).

You cannot use vector features on bitmap tile sources (see the [BitmapTileActivity](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/BitmapTileActivity.java) sample).

**Reply by firdauskz01** [reply]
Hi @devemux86 , thanks for your reply!
So, basically the map is need to be download it first if I wanna using Vector Tile?
my apologize i'm bit confused

**Reply by devemux86** [reply]
You need to study the [vtm-android-example](https://github.com/mapsforge/vtm/tree/master/vtm-android-example) for how to use the library.

There are samples with offline vector map files and online vector or bitmap tiles.

**Reply by firdauskz01** [reply]
I have run [vtm-android-example](https://github.com/mapsforge/vtm/tree/master/vtm-android-example). and on my phone (SDK 30), and tried using emulator (sdk 24) still blank if I run 3d Vector Building (Already see on [SimpleMapActivity](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/SimpleMapActivity.java)).
 see this 
![image](https://user-images.githubusercontent.com/19215383/122923283-c04caf80-d38e-11eb-860a-91ba6dfa9488.png) ![image](https://user-images.githubusercontent.com/19215383/122923400-e1150500-d38e-11eb-8bf0-800917429b96.png)

So why is this blank @devemux86 ?

**Reply by devemux86** [reply]
Perhaps that online vector tile provider does not work.

Try (online) MapilionMvtActivity, (online) BitmapTileActivity or (offline) Mapsforge.

**Reply by firdauskz01** [reply]
ok, so based on [this discussion](https://github.com/mapsforge/vtm/discussions/841), opensciencemap is does not work anymore? SO I can't use 3d buildings, right?

**Reply by devemux86** [reply]
You can use 3D buildings with other vector maps.
Offline Mapsforge maps v5 or online vector tiles.

**Reply by firdauskz01** [reply]
Where I can find other online vector tiles @devemux86 ?

**Reply by devemux86** [reply]
> Where I can find other online vector tiles

As mentioned above study the [vtm-android-example](https://github.com/mapsforge/vtm/tree/master/vtm-android-example).

There are samples with online vector tiles, like Mapilion or OpenMapTiles.
You can probably find more if you search the internet.

**Reply by firdauskz01** [reply]
Thanks @devemux86 
I apologize, I'm still studying it :)

Now I'm using the offline map, downloaded from [mapsforge v5](https://download.mapsforge.org/maps/v5)

and try using this code 
```
 File mapFile = getExternalFilesDir("indonesia.map");
  MultiMapFileTileSource tileSource = new MultiMapFileTileSource();
//            for (File mapFile : mapFiles) {
            MapFileTileSource mapFileTileSource = new MapFileTileSource();
            mapFileTileSource.setMapFile(mapFile.getAbsolutePath());
            tileSource.add(mapFileTileSource);
//            }

  VectorTileLayer tileLayer = mapView.map().setBaseMap(tileSource);

  // Building layer
            mapView.map().layers().add(new BuildingLayer(mapView.map(), tileLayer));

            // Label layer
            mapView.map().layers().add(new LabelLayer(mapView.map(), tileLayer));

            // Render theme
            mapView.map().setTheme(VtmThemes.DEFAULT);
```

but still shows blank. 

![image](https://user-images.githubusercontent.com/19215383/122936975-cd23d000-d39b-11eb-934a-526bb84d89c6.png)

**Reply by devemux86** [reply]
It is impossible to debug everyone's code.

There are the [getting started](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/GettingStarted.java) and other samples, you can study and compare your code with them.

**Reply by firdauskz01** [reply]
Yes agreed, thank you so much @devemux86 
now it's working with offline maps. Really apologize I'm still learning about it. :)
Now, I will search the online vector tiles.
![Screenshot_2021-06-22-21-40-27-72_4b474022c0f6272e0de1bfd1a9befee7](https://user-images.githubusercontent.com/19215383/122945294-b03ecb00-d3a2-11eb-8d30-73f289399c84.jpg)
